### PR TITLE
Add compatibilty with rack 2+

### PR DESCRIPTION
**Pull request by pabse**
This closes #124 #127

### CODE REVIEW

**Comment by atomless** [comment]
Is this fully working with Rack 2.0? And if so any idea why it has still not been merged into the master repo yet?

**Comment by mpalmer** [comment]
The failing CI is the immediate blocker.  Apart from that, a lack of time to sit down and merge outstanding PRs and cut a new release is where things are held up.

**Comment by pabse** [comment]
CI failed because rack 2+ requires ruby >= 2.2.2.
Fixed it 👍

**Comment by blueplanet** [comment]
@mpalmer Please merge this ! Thanks!

**Comment by rastating** [comment]
+1  to merge this :+1:

**Comment by takumi-tsuchiya-sms** [comment]
https://github.com/kickstarter/rack-attack

こいつでどうにかならないだろうか？？

**Comment by gonace** [comment]
+1 to merge this!

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,12 @@
 source 'https://rubygems.org'
 
 gemspec
+
```
Comment (mpalmer) [inline Gemfile]: Shouldn't these changes go in the gemspec?  Otherwise these restrictions will only be effective during development of rack-contrib, and not have any effect on installations.

File: lib/rack/contrib/jsonp.rb [hunk 2]
```diff
@@ -54,7 +54,9 @@ def call(env)
         
         # Set new Content-Length, if it was set before we mutated the response body
         if headers['Content-Length']
-          length = response.to_ary.inject(0) { |len, part| len + bytesize(part) }
+          length = response.to_ary.inject(0) do |len, part|
+            len + ( Rack.release < '2.0.0' ? bytesize(part) : part.bytesize )
```
Comment (mpalmer) [inline lib/rack/contrib/jsonp.rb]: Version strings can't be compared as strings; this will return `true` for `Rack.release = "10.0.0"`.  Use `Gem::Version.new(Rack.release) < Gem::Version.new("2.0.0")`.

File: lib/rack/contrib/sendfile.rb [hunk 3]
```diff
@@ -1,10 +1,6 @@
 require 'rack/file'
 
 module Rack
-  class File #:nodoc:
-    alias :to_path :path
```
Comment (mpalmer) [inline lib/rack/contrib/sendfile.rb]: Why was this removed?  Looks like vestigial code based on a quick glance, but I don't see how this is a Rack 2.x compatibility change.

File: rack-contrib.gemspec [hunk 4]
```diff
@@ -34,7 +34,7 @@ Gem::Specification.new do |s|
   # REMINDER: If you modify any dependencies, please ensure you
   # update `test/gemfiles/minimum_versions`!
   #
-  s.add_runtime_dependency 'rack', '~> 1.4'
+  s.add_runtime_dependency 'rack', '>= 1.4'
```
Comment (mpalmer) [inline rack-contrib.gemspec]: This would match Rack 3.0.0, which is far from guaranteed to be compatible (given it doesn't exist).

File: test/spec_rack_nested_params.rb [hunk 5]
```diff
@@ -1,7 +1,11 @@
 require 'minitest/autorun'
 require 'rack/mock'
 require 'rack/contrib/nested_params'
-require 'rack/methodoverride'
+if Rack.release < '2.0.0'
```
Comment (mpalmer) [inline test/spec_rack_nested_params.rb]: Another string-based version comparison here.

File: test/spec_rack_sendfile.rb [hunk 6]
```diff
@@ -2,9 +2,11 @@
 require 'rack/mock'
 require 'rack/contrib/sendfile'
 
-describe "Rack::File" do
-  specify "should respond to #to_path" do
-    Rack::File.new(Dir.pwd).must_respond_to :to_path
+if Rack.release < '2.0.0'
```
Comment (mpalmer) [inline test/spec_rack_sendfile.rb]: String-based version comparison.  Also, rummaging around, , and I get the feeling that `Rack::File`'s behaviour has changed in Rack 2.x, to the point where `Rack::Sendfile` *probably* won't work at all, I feel.

File: lib/rack/contrib/sendfile.rb (hunk 3 above)
Comment (tobiashm) [inline lib/rack/contrib/sendfile.rb]: `Rack::File` no longer has a `path` method in 2.0 – see e.g. https://github.com/rack/rack-contrib/issues/127#issuecomment-230248380

**Comment by tobiashm** [comment]
There seems to be a number of issues with supporting both Rack 1.x and 2.x at the same time.
Perhaps it would be an idea to prepare a `rack-contrib` 2.0 release that will support Rack 2.x only? And require you to use the old (current) release for anything using Rack 1.x

**Comment by mpalmer** [comment]
Yep, that's pretty much where I'm going with rack 2 support.  Clean break, require rack 2, ruby 2.2.2, break a few other backwards-compatibility things while we're there.  Unfortunately there's some stuff I'd really like to *deprecate* first, but I'm not sure how to do both at once...

**Comment by volkanunsal** [comment]
Any idea on when these changes might be merged?

**Comment by mpalmer** [comment]
Sometime after the changes that have been requested have been made.

**Comment by DannyBen** [comment]
So I understand that rack-contrib does not work with rack 2.0 (personally I am unable to upgrade to sinatra 2 now), and there is PR to fix it that is opened for over a year, and it is now harder to merge since a year has passed, and more changes flowed in. Is that correct?

If so, avoiding this merge will only make the problem worse. Is rack-contrib falling out of maintenance?

**Comment by mpalmer** [comment]
Rack-contrib is being maintained, insofar as PRs are reviewed and merged if deemed suitable.  If you would like to see a change in rack-contrib, please contribute a PR.  Complaining that nobody else has done work for you is... unproductive.

**Comment by sshaw** [comment]
I've made some updates based on this PR and ensuing conversation. See #127.

**Comment by mpalmer** [comment]
Closing this PR in favour of #137.
